{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import limix\n",
    "import qtl_output\n",
    "import qtl_loader_utils\n",
    "from qtl_snp_qc import do_snp_qc\n",
    "import glob\n",
    "import os\n",
    "from sklearn.preprocessing import Imputer\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# ##loader functions\n",
    "# def _ensure_dir(file_path):\n",
    "#     '''Check if directory exists for output, and create it if it doesn't.'''\n",
    "#     directory = os.path.dirname(file_path)\n",
    "#     if not os.path.exists(directory):\n",
    "#         os.makedirs(directory)\n",
    "\n",
    "# def _get_kinship_df(kinship_filename):\n",
    "#     if kinship_filename:\n",
    "#         kinship_df = pd.read_csv(kinship_filename,sep='\\t',index_col=0)\n",
    "#     else:\n",
    "#         kinship_df = None\n",
    "#     return kinship_df\n",
    "\n",
    "# def _get_samplemapping_df(sample_mapping_filename,sample_labels,key_from):\n",
    "#     assert(key_from in ['iid','sample'])\n",
    "#     if sample_mapping_filename:\n",
    "#         mapping_df = pd.read_csv(sample_mapping_filename,sep='\\t',header=None,names=['iid','sample'])\n",
    "#         mapping_df.set_index(key_from,inplace=True)\n",
    "#     else:\n",
    "#         #assume the mapping is the identity mapping\n",
    "#         identifiers = sample_labels\n",
    "#         mapping_df = pd.DataFrame(data=identifiers,index=identifiers,columns=['sample'])\n",
    "#     return mapping_df\n",
    "\n",
    "# def _get_covariate_df(covariates_filename):\n",
    "#     if covariates_filename:\n",
    "#         covariate_df = pd.read_csv(covariates_filename,sep='\\t',index_col=0)\n",
    "#     else:\n",
    "#         covariate_df = None\n",
    "#     return covariate_df\n",
    "\n",
    "# def _get_genotype_data(geno_prefix):\n",
    "#     bim,fam,bed = limix.io.read_plink(geno_prefix,verbose=False)\n",
    "#     fam.set_index('iid',inplace=True)\n",
    "#     return bim,fam,bed\n",
    "\n",
    "# def _get_annotation_df(anno_filename):\n",
    "#     annotation_col_dtypes = {'feature_id':np.object,\n",
    "#                          'gene_id':np.object,\n",
    "#                          'gene_name':np.object,\n",
    "#                          'chromosome':np.object,\n",
    "#                          'start':np.int64,\n",
    "#                          'end':np.int64,\n",
    "#                          'strand':np.object}\n",
    "#     annotation_df = pd.read_csv(anno_filename,sep='\\t',index_col=0,dtype=annotation_col_dtypes)\n",
    "#     return annotation_df\n",
    "\n",
    "# def _get_phenotype_df(pheno_filename):\n",
    "#     return pd.read_csv(pheno_filename,sep='\\t',index_col=0)\n",
    "\n",
    "# def do_snp_qc(snp_df, min_call_rate, min_maf, min_hwe_P):\n",
    "   \n",
    "#     #Determine call rate.\n",
    "#     call_rate = 1-snp_df.isnull().sum()/len(snp_df.index)\n",
    "#     #print(call_rate)\n",
    "#     #print(call_rate < min_call_rate)\n",
    "#     failed_snp_names  = list(snp_df.columns[call_rate < min_call_rate])\n",
    "#     #pass_qc_snps_all = list(snp_df.columns[call_rate >= min_call_rate])\n",
    "#     snp_df = snp_df.loc[:,list(snp_df.columns[call_rate >= min_call_rate])]\n",
    "\n",
    "#     #Determine MAF.\n",
    "#     genotypeCounter = np.zeros((len(snp_df.columns),3))\n",
    "#     for snp in range(0, len(snp_df.columns)):\n",
    "#         for sample in range(0, len(snp_df.index)):\n",
    "#             if not math.isnan(snp_df.iloc[sample,snp]) :\n",
    "#                 #print(snp_df.iloc[sample,snp])\n",
    "#                 #print(i)\n",
    "#                 #print(snp)\n",
    "#                 #print(int(round(snp_df.iloc[snp,i])))\n",
    "#                 genotypeCounter[snp,int(round(snp_df.iloc[sample,snp]))] +=1\n",
    "\n",
    "#     #Here we make sure that the major allele is temporarly \"coded\" as 0.\n",
    "#     for snp in range(0, len(snp_df.columns)):\n",
    "#         if genotypeCounter[snp,0]<genotypeCounter[snp,2]:\n",
    "#             tmp = genotypeCounter[snp,0]\n",
    "#             genotypeCounter[snp,0] = genotypeCounter[snp,2]\n",
    "#             genotypeCounter[snp,2] = tmp\n",
    "\n",
    "#     #print(genotypeCounter)\n",
    "#     maf = np.zeros((len(snp_df.columns)))\n",
    "#     for snp in range(0, len(snp_df.columns)):\n",
    "#         maf[snp] = ((genotypeCounter[snp,2]*2)+genotypeCounter[snp,1])/(len(snp_df.index)*2)\n",
    "\n",
    "#     failed_snp_names.extend(list(snp_df.columns[maf < min_maf]))\n",
    "#     snp_df = snp_df.loc[:,list(snp_df.columns[maf >= min_maf])]    \n",
    "#     genotypeCounter = genotypeCounter[maf >= min_maf,]\n",
    "\n",
    "#     #Determine HWE.\n",
    "#     hweP = np.zeros((len(snp_df.columns)))\n",
    "\n",
    "#     for snp in range(0, len(snp_df.columns)):\n",
    "#         rare_copies = int(((genotypeCounter[snp,2])*2)+genotypeCounter[snp,1])\n",
    "#         genotypes = (genotypeCounter[snp,2]+genotypeCounter[snp,1]+genotypeCounter[snp,0])\n",
    "#         het_probs = np.ones((rare_copies+1))\n",
    "#         # start at midpoint #\n",
    "\n",
    "#         mid = int(round(rare_copies * (2 * genotypes - rare_copies) / (2 * genotypes)))\n",
    "\n",
    "#         # check to ensure that midpoint and rare alleles have same parity #\n",
    "#         if mid % 2 is not rare_copies % 2 :\n",
    "#             mid+=1\n",
    "\n",
    "#         curr_homr = (rare_copies - mid) / 2;\n",
    "#         curr_homc = genotypes - mid - curr_homr;\n",
    "#         #print(mid)\n",
    "#         #print(het_probs.shape)\n",
    "#         sum_values = het_probs[mid];\n",
    "\n",
    "#         for curr_hets in range(mid, 2, -2) :\n",
    "#             het_probs[curr_hets - 2] = het_probs[curr_hets] * curr_hets * (curr_hets - 1.0) / (4.0 * (curr_homr + 1.0) * (curr_homc + 1.0))\n",
    "#             sum_values += het_probs[curr_hets - 2]\n",
    "#             ## 2 fewer heterozygotes for next iteration -> add one rare, one common homozygote ##\n",
    "#             curr_homr+=1\n",
    "#             curr_homc+=1\n",
    "\n",
    "\n",
    "#         curr_homr = (rare_copies - mid) / 2;\n",
    "#         curr_homc = genotypes - mid - curr_homr;\n",
    "#         for curr_hets in range(mid, (int(rare_copies) - 2), 2) :\n",
    "#             het_probs[curr_hets + 2] = het_probs[curr_hets] * 4.0 * curr_homr * curr_homc / ((curr_hets + 2.0) * (curr_hets + 1.0))\n",
    "#             sum_values += het_probs[curr_hets + 2]\n",
    "#             curr_homr-=1\n",
    "#             curr_homc-=1\n",
    "\n",
    "#         p_hwe = 0.0\n",
    "#         het_probs[int(genotypeCounter[snp,1])]/= sum_values\n",
    "#         for index in range(0,int(rare_copies)) :\n",
    "#             if index != int(genotypeCounter[snp,1]) :\n",
    "#                 het_probs[index] /= sum_values\n",
    "#             if het_probs[index] <= het_probs[int(genotypeCounter[snp,1])] :\n",
    "#                 p_hwe += het_probs[index];\n",
    "\n",
    "#         hweP[snp] = 1 if p_hwe > 1.0 else p_hwe;\n",
    "\n",
    "#     failed_snp_names.extend(list(snp_df.columns[hweP < min_hwe_P]))\n",
    "#     snp_df = snp_df.loc[:,list(snp_df.columns[hweP >= min_hwe_P])]\n",
    "\n",
    "#     return snp_df.columns, failed_snp_names\n",
    "\n",
    "def chunker(seq, size):\n",
    "    return (seq[pos:pos + size] for pos in range(0, len(seq), size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#settings\n",
    "pheno_filename = '/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/Expression/Geuvadis_CEU_YRI_Expr.txt.gz'\n",
    "anno_filename = '/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/Expression/Geuvadis_CEU_Annot.txt'\n",
    "geno_prefix = '/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/Genotypes/Geuvadis'\n",
    "window_size = 250000\n",
    "min_maf = 0.05\n",
    "min_hwe_P = 0.01\n",
    "min_call_rate = 0.95\n",
    "#cis_mode = True\n",
    "cis_mode = False\n",
    "\n",
    "output_dir = '/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/outputTest'\n",
    "chromosome='all'\n",
    "\n",
    "covariates_filename='/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/Expression/Geuvadis_CEU_YRI_covariates.txt'\n",
    "kinship_filename='/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/Genotypes/Geuvadis_chr1_kinship.txt'\n",
    "sample_mapping_filename='/hps/nobackup/stegle/users/mjbonder/tools/hipsci_pipeline/limix_QTL_pipeline/geuvadis_CEU_test_data/Geuvadis_CEU_gte.txt'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Core function to take input and run QTL tests on a given chromosome.\n",
    "#Load input data files & filter for relevant data\n",
    "bim,fam,bed = qtl_loader_utils.get_genotype_data(geno_prefix)\n",
    "phenotype_df = qtl_loader_utils.get_phenotype_df(pheno_filename)\n",
    "individual2sample_df = qtl_loader_utils.get_samplemapping_df(sample_mapping_filename,list(phenotype_df.columns),'iid')\n",
    "sample2individual_df = qtl_loader_utils.get_samplemapping_df(sample_mapping_filename,list(phenotype_df.columns),'sample')\n",
    "annotation_df = qtl_loader_utils.get_annotation_df(anno_filename)\n",
    "phenotype_df = phenotype_df.loc[annotation_df.index.values,individual2sample_df.loc[list(set(fam.index)&set(individual2sample_df.index)),'sample'].values]\n",
    "\n",
    "kinship_df = qtl_loader_utils.get_kinship_df(kinship_filename)    \n",
    "kinship_df = kinship_df.loc[list(set(fam.index)&set(individual2sample_df.index)),list(set(fam.index)&set(individual2sample_df.index))]\n",
    "\n",
    "covariate_df = qtl_loader_utils.get_covariate_df(covariates_filename)\n",
    "covariate_df = covariate_df.loc[individual2sample_df.loc[list(set(fam.index)&set(individual2sample_df.index)),'sample'].values,]\n",
    "\n",
    "### make sure covariates, kinship & phenotype is not changed in a later stage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000215014\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if(cis_mode):\n",
    "    #Remove features from the annotation that are on chromosomes which are not present anyway.\n",
    "    annotation_df = annotation_df[annotation_df['chromosome'].map(lambda x: x in list(set(bim['chrom'])))]\n",
    "    annotation_df = annotation_df[annotation_df['chromosome'].map(lambda x: x in list(map(str, range(1, 23))))]\n",
    "\n",
    "#Determine features to be tested\n",
    "if chromosome=='all':\n",
    "    feature_list = list(set(annotation_df.index)&set(phenotype_df.index))    \n",
    "else:\n",
    "    feature_list = list(set(annotation_df[annotation_df['chromosome']==chromosome].index)&set(phenotype_df.index))\n",
    "\n",
    "#Array to store indices of snps tested\n",
    "tested_snp_idxs = []\n",
    "\n",
    "feature_id = feature_list[1]\n",
    "print(feature_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "1510355\n",
      "1511373\n",
      "   chrom          snp   cm    pos a0 a1  i\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "1      1  snp_1_82676  0.0  82676  G  T  1\n",
      "..   ...          ...  ...    ... .. .. ..\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "0      1  snp_1_79772  0.0  79772  G  C  0\n",
      "\n",
      "[3151 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "chrom = str(annotation_df.loc[feature_id,'chromosome'])\n",
    "print(chrom)\n",
    "start = annotation_df.loc[feature_id,'start']\n",
    "print(start)\n",
    "end = annotation_df.loc[feature_id,'end']\n",
    "print(end)\n",
    "\n",
    "#make robust to features specified back-to-front\n",
    "lowest = min([start,end])\n",
    "highest = max([start,end])\n",
    "\n",
    "if (cis_mode) : \n",
    "    snpQuery = bim.query(\"chrom == '%s' & pos > %d & pos < %d\" % (chrom, lowest-window_size, highest+window_size))\n",
    "else :\n",
    "    snpQuery = bim.query(\"(chrom == '%s' & (pos < %d | pos > %d))|chrom != '%s'\" % (chrom, lowest-window_size, highest+window_size,chrom))\n",
    "    snpQuery = snpQuery.iloc[snpQuery['chrom'].map(lambda x: x in list(map(str, range(1, 23))))]\n",
    "    \n",
    "# blocksize = 500\n",
    "\n",
    "# print(len(snpQuery))\n",
    "# if len(snpQuery) != 0:\n",
    "\n",
    "#     print(len(snpQuery)%blocksize)\n",
    "#     previous =0\n",
    "#     results_df = pd.DataFrame()\n",
    "#     print(results_df.shape)\n",
    "#     for snpGroup in chunker(snpQuery, blocksize):\n",
    "#         #Here we need to batch for cis & trans.\n",
    "#         snp_idxs = snpGroup['i'].values\n",
    "#         snp_names = snpGroup['snp'].values\n",
    "\n",
    "#         tested_snp_idxs.extend(snp_idxs)\n",
    "\n",
    "#         ##Check for NA's in feature. Remove samples with NA's\n",
    "\n",
    "#         sample_ids = individual2sample_df.loc[:,'sample'].values\n",
    "\n",
    "#         phenotype_ds = phenotype_df.loc[feature_id,sample_ids]\n",
    "#         contains_missing_samples = any(phenotype_ds.isnull().values)\n",
    "#         phenotype_ds.dropna(inplace=True)\n",
    "\n",
    "#         #indices for relevant individuals in genotype matrix    \n",
    "#         individual_ids = list(set(fam.index)&set(sample2individual_df.loc[phenotype_ds.index,'iid']))\n",
    "#         individual_idxs = fam.loc[individual_ids,'i'].values\n",
    "\n",
    "#         #subset genotype matrix, we cannot subselect at the same time, do in two steps.\n",
    "#         snp_df = pd.DataFrame(data=bed[snp_idxs,:].compute().transpose(),index=fam.index,columns=snp_names)\n",
    "#         snp_df = snp_df.loc[individual_ids,:]\n",
    "#         print(snp_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(90, 151)\n",
      "False\n",
      "          feature_id         snp_id   p_value      beta  n_samples\n",
      "0    ENSG00000215014  snp_2_8980472  0.247541  0.118171         90\n",
      "1    ENSG00000215014  snp_2_8988660  0.561975  0.085046         90\n",
      "2    ENSG00000215014  snp_2_8996746  0.561975  0.085046         90\n",
      "3    ENSG00000215014  snp_2_9007398  0.775673  0.045281         90\n",
      "4    ENSG00000215014  snp_2_9041284  0.894102  0.021493         90\n",
      "5    ENSG00000215014  snp_2_9061326  0.557266  0.104928         90\n",
      "6    ENSG00000215014  snp_2_9064878  0.894102  0.021493         90\n",
      "7    ENSG00000215014  snp_2_9071427  0.697416  0.059178         90\n",
      "8    ENSG00000215014  snp_2_9082673  0.557266  0.104928         90\n",
      "9    ENSG00000215014  snp_2_9094995  0.894102  0.021493         90\n",
      "10   ENSG00000215014  snp_2_9167124  0.965390  0.007989         90\n",
      "11   ENSG00000215014  snp_2_9228515  0.369972  0.102101         90\n",
      "12   ENSG00000215014  snp_2_9233610  0.764660 -0.037929         90\n",
      "13   ENSG00000215014  snp_2_9239563  0.647314 -0.058709         90\n",
      "14   ENSG00000215014  snp_2_9241588  0.647314 -0.058709         90\n",
      "15   ENSG00000215014  snp_2_9247782  0.931769  0.013223         90\n",
      "16   ENSG00000215014  snp_2_9257378  0.142601  0.146825         90\n",
      "17   ENSG00000215014  snp_2_9263253  0.015511  0.293380         90\n",
      "18   ENSG00000215014  snp_2_9290357  0.292634  0.106090         90\n",
      "19   ENSG00000215014  snp_2_9295417  0.195267 -0.135212         90\n",
      "20   ENSG00000215014  snp_2_9296371  0.263162 -0.117102         90\n",
      "21   ENSG00000215014  snp_2_9305035  0.572071 -0.063023         90\n",
      "22   ENSG00000215014  snp_2_9310102  0.393623 -0.095803         90\n",
      "23   ENSG00000215014  snp_2_9327543  0.534730 -0.085527         90\n",
      "24   ENSG00000215014  snp_2_9344759  0.299906 -0.133625         90\n",
      "25   ENSG00000215014  snp_2_9355318  0.447720 -0.077919         90\n",
      "26   ENSG00000215014  snp_2_9376835  0.256843  0.140247         90\n",
      "27   ENSG00000215014  snp_2_9390392  0.055812 -0.234166         90\n",
      "28   ENSG00000215014  snp_2_9393303  0.689274  0.041146         90\n",
      "29   ENSG00000215014  snp_2_9409017  0.055812 -0.234166         90\n",
      "..               ...            ...       ...       ...        ...\n",
      "72   ENSG00000215014  snp_2_9822136  0.991234 -0.001235         90\n",
      "73   ENSG00000215014  snp_2_9831821  0.229982 -0.144527         90\n",
      "74   ENSG00000215014  snp_2_9834465  0.095855 -0.302250         90\n",
      "75   ENSG00000215014  snp_2_9860103  0.595450 -0.069610         90\n",
      "76   ENSG00000215014  snp_2_9865448  0.991832  0.001121         90\n",
      "77   ENSG00000215014  snp_2_9865924  0.567132  0.079020         90\n",
      "78   ENSG00000215014  snp_2_9872385  0.587874 -0.068295         90\n",
      "79   ENSG00000215014  snp_2_9880217  0.335454  0.123021         90\n",
      "80   ENSG00000215014  snp_2_9888428  0.617775  0.066856         90\n",
      "81   ENSG00000215014  snp_2_9889167  0.617775  0.066856         90\n",
      "82   ENSG00000215014  snp_2_9890717  0.617775  0.066856         90\n",
      "83   ENSG00000215014  snp_2_9904601  0.270676 -0.198435         90\n",
      "84   ENSG00000215014  snp_2_9907364  0.247652 -0.160007         90\n",
      "85   ENSG00000215014  snp_2_9913779  0.877392  0.017283         90\n",
      "86   ENSG00000215014  snp_2_9922073  0.316479 -0.138145         90\n",
      "87   ENSG00000215014  snp_2_9925449  0.469942 -0.100078         90\n",
      "88   ENSG00000215014  snp_2_9933427  0.021641 -0.230259         90\n",
      "89   ENSG00000215014  snp_2_9936204  0.022660 -0.268749         90\n",
      "90   ENSG00000215014  snp_2_9937942  0.992744 -0.001155         90\n",
      "91   ENSG00000215014  snp_2_9937945  0.152513 -0.157988         90\n",
      "92   ENSG00000215014  snp_2_9956861  0.066596 -0.178198         90\n",
      "93   ENSG00000215014  snp_2_9966470  0.139843  0.206482         90\n",
      "94   ENSG00000215014  snp_2_9968344  0.139843  0.206482         90\n",
      "95   ENSG00000215014  snp_2_9972442  0.139843  0.206482         90\n",
      "96   ENSG00000215014  snp_2_9975557  0.280689  0.185483         90\n",
      "97   ENSG00000215014  snp_2_9978026  0.704504  0.064957         90\n",
      "98   ENSG00000215014  snp_2_9984829  0.828337 -0.047724         90\n",
      "99   ENSG00000215014  snp_2_9985606  0.361124  0.106326         90\n",
      "100  ENSG00000215014  snp_2_9987675  0.393754 -0.099193         90\n",
      "101  ENSG00000215014  snp_2_9989951  0.405989 -0.108878         90\n",
      "\n",
      "[102 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "pass_qc_snps_all = []\n",
    "fail_qc_snps_all = []\n",
    "print(snp_df.shape)\n",
    "print(contains_missing_samples)\n",
    "\n",
    "#For the updated QC where we only use one individual per genotype. We can just select sites from df to test.\n",
    "if not contains_missing_samples:\n",
    "    #remove snps from snp_df if they fail QC\n",
    "    snp_df = snp_df.loc[:,snp_df.columns[~snp_df.columns.isin(fail_qc_snps_all)]]\n",
    "    snps_to_test_df = snp_df.loc[:,snp_df.columns[~snp_df.columns.isin(pass_qc_snps_all)]]\n",
    "    \n",
    "    #Only do QC on relevant SNPs. join pre-QCed list and new QCed list.\n",
    "    #passed_snp_names,failed_snp_names = snp_qc.do_snp_qc(snps_to_test_df)\n",
    "    #File will be called snp_qc.py\n",
    "    passed_snp_names,failed_snp_names = do_snp_qc(snps_to_test_df, min_call_rate, min_maf, min_hwe_P)\n",
    "    snps_to_test_df = None\n",
    "    \n",
    "    #append snp_names and failed_snp_names\n",
    "    pass_qc_snps_all.extend(passed_snp_names)\n",
    "    fail_qc_snps_all.extend(failed_snp_names)\n",
    "    \n",
    "    snp_names  = list(snp_df.columns[snp_df.columns.isin(pass_qc_snps_all)])\n",
    "    #snp_names = ['snp_1_2739933', 'snp_1_2750437', 'snp_1_2752719', 'snp_1_2753372']\n",
    "    snp_df = snp_df.loc[:,snp_df.columns[snp_df.columns.isin(pass_qc_snps_all)]]\n",
    "\n",
    "else:\n",
    "    #Do snp QC for relevant section.\n",
    "    passed_snp_names,failed_snp_names = do_snp_qc(snp_df, min_call_rate, min_maf, min_hwe_P)\n",
    "    snp_df = snp_df.loc[:,snp_df.columns[snp_df.columns.isin(pass_qc_snps_all)]]\n",
    "    \n",
    "    \n",
    "# if len(snp_df.columns) != 0:\n",
    "#     continue\n",
    "\n",
    "snp_matrix = snp_df.values\n",
    "snp_df = None\n",
    "\n",
    "snp_matrix = Imputer(missing_values=np.nan,strategy='mean',axis=0,copy=False).fit_transform(snp_matrix)\n",
    "\n",
    "if kinship_df is not None:\n",
    "    kinship_mat = kinship_df.loc[individual_ids,individual_ids].values\n",
    "else:\n",
    "    kinship_mat = None\n",
    "\n",
    "#map individual_ids to samples\n",
    "sample_ids = individual2sample_df.loc[individual_ids,'sample'].values\n",
    "phenotype = phenotype_ds.loc[sample_ids].values\n",
    "\n",
    "#generate covariate matrix\n",
    "if covariate_df is not None:\n",
    "    cov_matrix = np.concatenate([np.ones((len(sample_ids),1)),covariate_df.loc[sample_ids,:].values],axis=1)\n",
    "else:\n",
    "    cov_matrix = None\n",
    "\n",
    "#fit modelrun\n",
    "LMM = limix.qtl.qtl_test_lmm(snp_matrix, phenotype,K=kinship_mat,covs=cov_matrix)\n",
    "\n",
    "#add these results to qtl_results\n",
    "\n",
    "temp_df = pd.DataFrame(index = range(len(snp_names)),columns=['feature_id','snp_id','p_value','beta','n_samples'])\n",
    "temp_df['snp_id'] = snp_names\n",
    "temp_df['feature_id'] = feature_id\n",
    "temp_df['beta'] = LMM.getBetaSNP()[0]\n",
    "temp_df['p_value'] = LMM.getPv()[0]\n",
    "temp_df['n_samples'] = sum(~np.isnan(phenotype))\n",
    "results_df = results_df.append(temp_df)\n",
    "print(temp_df)\n",
    "# output_writer.add_result_df(temp_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         feature_id         snp_id   p_value      beta  n_samples\n",
      "0   ENSG00000215014  snp_1_1276077  0.933442  0.009569         90\n",
      "1   ENSG00000215014  snp_1_1287922  0.027078 -0.523502         90\n",
      "2   ENSG00000215014  snp_1_1299633  0.326317  0.206344         90\n",
      "3   ENSG00000215014  snp_1_1308387  0.326317  0.206344         90\n",
      "4   ENSG00000215014  snp_1_1469782  0.135740 -0.371984         90\n",
      "5   ENSG00000215014  snp_1_1526799  0.198926  0.215296         90\n",
      "6   ENSG00000215014  snp_1_1527952  0.087492  0.210162         90\n",
      "7   ENSG00000215014  snp_1_1535700  0.357051  0.118735         90\n",
      "8   ENSG00000215014  snp_1_1545289  0.215915  0.160522         90\n",
      "9   ENSG00000215014  snp_1_1562536  0.420513 -0.140472         90\n",
      "10  ENSG00000215014  snp_1_1584102  0.826237  0.025858         90\n",
      "11  ENSG00000215014  snp_1_1590681  0.936862  0.009429         90\n",
      "12  ENSG00000215014  snp_1_1602536  0.835972 -0.025113         90\n",
      "13  ENSG00000215014  snp_1_1615348  0.871757 -0.019827         90\n",
      "14  ENSG00000215014  snp_1_1616058  0.925313 -0.011710         90\n",
      "15  ENSG00000215014  snp_1_1635619  0.942291  0.009637         90\n",
      "16  ENSG00000215014  snp_1_1658343  0.805616  0.029125         90\n",
      "17  ENSG00000215014  snp_1_1665702  0.114955 -0.232309         90\n",
      "18  ENSG00000215014  snp_1_1686040  0.298082 -0.154525         90\n",
      "19  ENSG00000215014  snp_1_1693726  0.976925 -0.004498         90\n",
      "20  ENSG00000215014  snp_1_1699201  0.596849 -0.083144         90\n",
      "21  ENSG00000215014  snp_1_1703303  0.976925 -0.004498         90\n",
      "22  ENSG00000215014  snp_1_1704899  0.976925 -0.004498         90\n",
      "23  ENSG00000215014  snp_1_1722932  0.976925 -0.004498         90\n",
      "0   ENSG00000215014  snp_1_1276077  0.933442  0.009569         90\n",
      "1   ENSG00000215014  snp_1_1287922  0.027078 -0.523502         90\n",
      "2   ENSG00000215014  snp_1_1299633  0.326317  0.206344         90\n",
      "3   ENSG00000215014  snp_1_1308387  0.326317  0.206344         90\n",
      "4   ENSG00000215014  snp_1_1469782  0.135740 -0.371984         90\n",
      "5   ENSG00000215014  snp_1_1526799  0.198926  0.215296         90\n",
      "6   ENSG00000215014  snp_1_1527952  0.087492  0.210162         90\n",
      "7   ENSG00000215014  snp_1_1535700  0.357051  0.118735         90\n",
      "8   ENSG00000215014  snp_1_1545289  0.215915  0.160522         90\n",
      "9   ENSG00000215014  snp_1_1562536  0.420513 -0.140472         90\n",
      "10  ENSG00000215014  snp_1_1584102  0.826237  0.025858         90\n",
      "11  ENSG00000215014  snp_1_1590681  0.936862  0.009429         90\n",
      "12  ENSG00000215014  snp_1_1602536  0.835972 -0.025113         90\n",
      "13  ENSG00000215014  snp_1_1615348  0.871757 -0.019827         90\n",
      "14  ENSG00000215014  snp_1_1616058  0.925313 -0.011710         90\n",
      "15  ENSG00000215014  snp_1_1635619  0.942291  0.009637         90\n",
      "16  ENSG00000215014  snp_1_1658343  0.805616  0.029125         90\n",
      "17  ENSG00000215014  snp_1_1665702  0.114955 -0.232309         90\n",
      "18  ENSG00000215014  snp_1_1686040  0.298082 -0.154525         90\n",
      "19  ENSG00000215014  snp_1_1693726  0.976925 -0.004498         90\n",
      "20  ENSG00000215014  snp_1_1699201  0.596849 -0.083144         90\n",
      "21  ENSG00000215014  snp_1_1703303  0.976925 -0.004498         90\n",
      "22  ENSG00000215014  snp_1_1704899  0.976925 -0.004498         90\n",
      "23  ENSG00000215014  snp_1_1722932  0.976925 -0.004498         90\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df = results_df.append(temp_df)\n",
    "print(results_df)\n",
    "results_df.empty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
